### ISSUE TITLE
Expose capabilities/requirements from dart:ui for text edit size/transform

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#46896]
Today, only Flutter Web needs to know about the size and transform of editable text widgets. However, we calculate it for all platforms and send a method up.  

One approach is to guard this with `kIsWeb`, which we would like to avoid.

Another approach (suggested in https://github.com/flutter/flutter/pull/46843#discussion_r357264257) would be to expose the requirement for this from dart:ui so the framework could more cleanly check if it needs to do this work.

Yet another approach would be to move the entire method channel interface into dart:ui, which should be treated separately (and may not be desireable due to introducing more compelxity/weight to dart:ui, including all the codec/channel logic that currently lives in the framework).

/cc @yjbanov  @goderbauer @nturgut @mdebbar

**Comment by yjbanov** [comment]
> Yet another approach would be to move the entire method channel interface into dart:ui, which should be treated separately (and may not be desireable due to introducing more compelxity/weight to dart:ui, including all the codec/channel logic that currently lives in the framework).

I do not believe it introduces complexity or weight. On the contrary, we would be able to use normal Dart code and dartdocs to enforce and document the contract between the engine and the framework. Currently this contract is implicit in the messages sent through method channels and it's not statically enforced.

**Comment by dnfield** [comment]
I had previously asked @hixie about why we don't move the Flutter standard codec related classes into the engine, which would simplify some test code we have - that was the answer I was given.

It's not just about moving the channel declarations, it's all the other machinery around them as well.

**Comment by Hixie** [comment]
The goal is for dart:ui to be as small as possible, with as much of the logic as possible on the Dart side so that people can replace it if they want. So dart:ui itself should be as unopinionated as possible.

**Comment by yjbanov** [comment]
@Hixie I think that's a good goal. However, moving text editing protocol into a method channel does not reduce the API surface of dart:ui, only obscures it. The `flutter/textinput` channel protocol is mandatory for dart:ui to implement. It's just not expressed via Dart, and therefore is not statically enforced, leading to runtime bugs. How would you prefer to deal with this problem?

**Comment by goderbauer** [comment]
> The flutter/textinput channel protocol is mandatory for dart:ui to implement

Why is textinput protocol mandatory for dart:ui? In theory you could write an embedder without text input e.g. for a display-only use case?

**Comment by mdebbar** [comment]
> Why is textinput protocol mandatory for dart:ui?

It's mandatory in the sense that dart:ui ***will*** receive it and has to handle it. It can choose to ignore it, sure.

**Comment by goderbauer** [comment]
I still don't understand? Currently, as far as I know text input is fully handled in the text input plugins. How does dart:ui come into this?

**Comment by mdebbar** [comment]
All platform messages go through `dart:ui`'s `window.sendPlatformMessage()` and are handled by the accompanying engine implementation.

Maybe text input isn't a good example, since it's an optional method channel. A better example might be `flutter/platform_views` or `flutter/navigation` that are not optional.

**Comment by Hixie** [comment]
In what sense are they not optional?
